### PR TITLE
*: fix executing shard DDL with mismatched request

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

1. DM-worker does not verify the `ExecDDLRequest` from the DM-master
2. DM-master may send duplicate `ExecDDLRequest` to DM-worker (in order to support re-fetching DDL info from DM-worker after DM-master restarted)

example:
1. execute `ADD INDEX idx1`
2. duplicate `ADD INDEX idx1` request sent
3. `ADD COLUMN c1` synced in DM-worker
4. duplicate request used, `ADD COLUMN c1` executed, but `ADD COLUMN c1` in other DM-workers not synced yet.

ref: https://internal.pingcap.net/jira/browse/ONCALL-745

### What is changed and how it works?

1. do not re-send DDL execution request in most scenarios
    1. ignore the `invalid connection` for `ADD INDEX`
    2. decrease DDL DB execution timeout, increase DDL gRPC execution request
2. ignore duplicate DDL execution request
    - verify DDL if possible

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

**Comment by csuzhangxc** [comment]
/run-compatibility-test case=all pre_commit=e63c6cdebea0edcf2ef8c91d84cff4aaa5fc2df7

**Comment by csuzhangxc** [comment]
@amyangfei @WangXiangUSTC PTAL

**Comment by csuzhangxc** [comment]
/run-compatibility-test case=all pre_commit=e63c6cdebea0edcf2ef8c91d84cff4aaa5fc2df7

File: syncer/error.go [hunk 1]
```diff
@@ -70,3 +75,53 @@ func originError(err error) error {
...
+			return err
+		} else if v.Specs[0].Tp == ast.AlterTableAddConstraint {
+			handle()
+			return nil // ignore the error
+		}
+	case *ast.CreateIndexStmt:
+		handle()
+		return nil // ignore the error
+	default:
+		return err
```
Comment (amyangfei) [inline syncer/error.go]: can remove the default case

File: syncer/error.go [hunk 2]
```diff
@@ -70,3 +75,53 @@ func originError(err error) error {
...
+// handleSpecialDDLError handles special errors for DDL execution.
+// it only ignore `invalid connection` error (timeout or other causes) for `ADD INDEX` now.
+// `invalid connection` means some data already sent to the server,
+// and we assume that the whole SQL statement has already sent to the server for this error.
+// if we have other methods to judge the DDL dispatched but timeout for executing, we can update this method.
+// NOTE: we must ensure other PK/UK exists for correctness.
+// NOTE: when we are refactoring the shard DDL algorithm, we also need to consider supporting non-blocking `ADD INDEX`.
+func (s *Syncer) handleSpecialDDLError(tctx *tcontext.Context, err error, ddls []string, index int, conn *DBConn) error {
+	// must ensure only the last statement executed failed with the `invalid connection` error
+	if len(ddls) == 0 || index != len(ddls)-1 || errors.Cause(err) != mysql.ErrInvalidConn {
```
Comment (amyangfei) [inline syncer/error.go]: Have a question for the `failed SQL index`, the underlying implementation in `ExecuteSQLWithIgnoreError`, if failed in the `commit` procedure, the returned index is `len(ddls) - 1`. Such as the DDLs are ``alter table t add column col int, add index col(`col`)``, what happens to the transaction if commit failed?

Comment (csuzhangxc) [reply syncer/error.go]: `begin` and `commit` has no effect on DDL, and DDL is always executed before `commit`.

File: syncer/error.go (hunk 1 above)
Comment (csuzhangxc) [reply syncer/error.go]: removed in https://github.com/pingcap/dm/pull/336/commits/81f4abb17f9fc03eed09a467aab6030ff1ffc805.

File: syncer/error.go [hunk 3]
```diff
@@ -70,3 +75,53 @@ func originError(err error) error {
...
+	}
+
+	switch v := stmt.(type) {
+	case *ast.AlterTableStmt:
+		// ddls should be split with only one spec
+		if len(v.Specs) > 1 {
+			return err
+		} else if v.Specs[0].Tp == ast.AlterTableAddConstraint {
+			handle()
+			return nil // ignore the error
```
Comment (WangXiangUSTC) [inline syncer/error.go]: why ignore error when tp is AlterTableAddConstraint?

Comment (csuzhangxc) [reply syncer/error.go]: `ast.AlterTableAddConstraint` represents `ALTER TABLE xxx ADD INDEX yyy`, and `ast.CreateIndexStmt` represents `CREATE INDEX yyy ON xxx`.

Comment (WangXiangUSTC) [reply syncer/error.go]: it may be `add column` or other type?

Comment (csuzhangxc) [reply syncer/error.go]: It only can be index and foreign key now in TiDB (https://github.com/pingcap/tidb/blob/3cdea0dfdf28197ee65545debce8c99e6d2945e3/ddl/ddl_api.go#L1929-L1948).

do you think also need to check `spec.Constraint.Tp` in DM?

Comment (WangXiangUSTC) [reply syncer/error.go]: ok, LGTM

Comment (csuzhangxc) [reply syncer/error.go]: add more check in https://github.com/pingcap/dm/pull/336/commits/e4b29b44890b82a951852e0ad33ee14e727087b7.

**Pull request merged**